[16.0] [Add] fs_attachment_s3: support X-Accel-Redirect for S3-backed attachments

lmignon opened the pull request:
Add fs_attachment_s3: support X-Accel-Redirect for authenticated S3 storage

This PR introduces a new addon, fs_attachment_s3, which simplifies the use of the X-Accel-Redirect mechanism to serve attachments stored on S3-compatible backends.

It enables X-Accel-Redirect even when attachments are stored on authenticated S3-like servers. This is achieved by having Odoo generate a signed internal URL that the reverse proxy (e.g., Nginx or Apache) can use to fetch and serve the file securely, without exposing credentials.

Example: Nginx configuration
If you're using Nginx as a reverse proxy and have enabled the "X-Sendfile to serve internal URL" option on your fs.storage configured for S3, you only need to add the following directive to your Nginx configuration. After that, Odoo will no longer serve the file content—Nginx will serve it directly:

```nginx
location ~ ^/fs_x_sendfile/(.*?)/(.*?)/(.*) {
    internal;

    set $url_scheme $1;
    set $url_host $2;
    set $url_path $3;
    set $url $url_scheme://$url_host/$url_path;

    proxy_pass $url$is_args$args;
    proxy_set_header Host $url_host;
    proxy_ssl_server_name on;
}
```

This improves performance by offloading file transfer to the reverse proxy, while maintaining secure access through signed URLs generated by Odoo in the case of private S3.

lmignon commented:
FYI @sbidoul 

I should maybe create a specific addon for the code specific to s3....

sbidoul commented on fs_attachment/models/ir_attachment.py:
```diff
@@ -677,6 +678,13 @@ def _get_storage_codes(self):
+        """Get the path to use for X-Accel-Redirect"""
+        self.ensure_one()
+        if not self.fs_url_path:
+            raise UserError(_("The attachment is not stored in a filesystem storage."))
```
That's probably not a user mistake?

sbidoul commented on fs_attachment/models/fs_storage.py:
```diff
@@ -501,3 +505,39 @@ def recompute_urls(self) -> None:
+            raise RuntimeError(
+                "The attachment %s is not stored in a filesystem storage."
+                % attachment.id
+            )
```
This error should come from the super() call?

sbidoul commented on fs_attachment/models/fs_storage.py:
```diff
@@ -501,3 +505,39 @@ def recompute_urls(self) -> None:
+            endpoint_url = s3_client.meta.endpoint_url
+            path /= url.replace(endpoint_url, "").lstrip("/")
+        else:
+            path /= attachment.fs_url_path.lstrip("/")
```
call super?

sbidoul reviewed (commented):
> I should maybe create a specific addon for the code specific to s3....

That's probably cleaner indeed.

sbidoul commented on fs_attachment/models/fs_storage.py:
```diff
@@ -501,3 +505,39 @@ def recompute_urls(self) -> None:
+            # exclude the endpoint_url from the path
+            # for s3 storage, we use a signed URL
+            endpoint_url = s3_client.meta.endpoint_url
+            path /= url.replace(endpoint_url, "").lstrip("/")
```
We may need a specific prefix here so the nginx config can define an *internal* `location` that proxies the S3 endpoint.

lmignon replied:
> We may need a specific prefix here so the nginx config can define an _internal_ `location` that proxies the S3 endpoint.

The prefix is the storage_code `path = Path("/") / attachment.fs_storage_code`

lmignon committed:
[FIX] fs_attachment: Simplify code

The implementation of the method for calculating the URL for accessing attachments from an external file system was unnecessarily complex, mainly due to a bug in the fsspec library, which has since been fixed from fspesc>=2025.3.0

lmignon committed:
[FIX] fs_attachment: Makes the calculation of the x-accel-redirect path overridable.

Depending on the type of file system used, it may be necessary to adapt the way the path is calculated for file access via x-accel-redirect. A good example is the need to generate pre-signed URLs to allow unauthenticated access to files stored in an S3  protected by authentication.

lmignon committed:
[FIX] fs_attachment_s3: Specific features for fs_attachment into S3

This module extends the functionality of [fs_attachment](https://github.com/OCA/storage/tree/16.0/fs_attachment)
to better support Amazon S3 storage. It includes features such as signed URLs for X-Accel-Redirect, and enforcing
the mimetype of files stored in S3.

sbidoul commented on fs_attachment_s3/README.rst:
```diff
@@ -0,0 +1,121 @@
+
+- **S3 Uses Signed URL For X-Accel-Redirect**: If checked, the
+  X-Accel-Redirect path will be a signed URL, which is useful for S3
+  storages that require signed URLs for access.
```
Worth explaining how the returned path is structured.

sbidoul commented on fs_attachment_s3/models/fs_storage.py:
```diff
@@ -0,0 +1,82 @@
+            # exclude the endpoint_url from the path
+            endpoint_url = s3_client.meta.endpoint_url
+            url_path = url.replace(endpoint_url, "").lstrip("/")
+            path = Path("/") / storage_code / url_path
```
I'm not entirely convinced the storage_code is that useful here, but to be seen when we integrate with the deployment pipeline.

lmignon replied:
The mechanism is now generic. All the storages use the same rule into nginx. 
```
location ~ ^/fs_x_sendfile/(.*?)/(.*?)/(.*) {
        internal;
        proxy_pass http://myserver.com;
        set $url_protocol $1;
        set $url_host $2;
        set $url_path $3;
        set $url $url_protocol://$url_host/$url_path;
        proxy_pass $url$is_args$args;
        proxy_set_header Host $url_host;
        proxy_ssl_server_name on;
     
        # Optional: Forward original filename (if needed)
        proxy_set_header X-Original-URI $request_uri;
    }
```

lmignon committed:
[IMP] fs_attachment: Improve X-Sendfile configuration handling

* Decouples the use of the X-Sendfile mechanism for externally stored content from the global X-Sendfile setting in Odoo.

* Previously, X-Sendfile had to be enabled both globally in Odoo and on each filesystem storage.

* With this change, Odoo no longer needs to have X-Sendfile enabled globally to allow its use at the external filesystem level.

lmignon committed:
[IMP] fs_attachment, fs_attachment_s3: x-send-file: Introduce generic management mechanism

Previously, enabling x-send-file required defining a dedicated redirection rule in the reverse proxy for each storage backend configured to use it.

This change introduces a generic approach that allows configuring a single reverse proxy rule applicable to all storages. The path used in X-Accel-Redirect now follows a standardized format:

The reverse proxy can parse this path to reconstruct the target URL and serve the file accordingly. This simplifies configuration and improves maintainability.

lmignon committed:
[IMP] fs_attachment: Restore original format for x-accel-redirect

The new variable “x_send_file_base_url” has also been removed to avoid unnecessarily complicating the code without any real commercial necessity.

lmignon committed:
[IMP] fs_attachement_s3: Simplfy x-accel-redirect path generation

sbidoul commented on fs_attachment_s3/models/fs_storage.py:
```diff
@@ -0,0 +1,114 @@
+        return fields
+
+    @api.model
+    def _get_x_accel_redirect_path(self, attachment: IrAttachment):
```
Why not a method on ir.attachment?

lmignon replied:
> Why not a method on ir.attachment?

I'm wondering the same thing.... so this method and the new one from the last commit must be moved on ir.attachment

sbidoul commented on fs_attachment_s3/readme/CONFIGURE.md:
```diff
@@ -0,0 +1,57 @@
+
+    location ~ ^/fs_x_accel_redirect/(.*?)/(.*?)/(.*) {
+        internal;
+        set $url_protocol $1;
```
```suggestion
        set $url_scheme $1;
```

sbidoul commented on fs_attachment_s3/readme/CONFIGURE.md:
```diff
@@ -0,0 +1,57 @@
+        set $url_protocol $1;
+        set $url_host $2;
+        set $url_path $3;
+        set $url $url_protocol://$url_host/$url_path;
```
```suggestion
        set $url $url_scheme://$url_host/$url_path;
```

sbidoul commented on fs_attachment_s3/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+    "depends": ["fs_attachment"],
+    "external_dependencies": {
+        "python": [
+            "s3fs",
```
```suggestion
            "fsspec[s3]",
```

Would this work?

lmignon replied:
AFAIK it should not work due to the way Odoo validates the declared python's external dependencies

sbidoul commented on fs_attachment_s3/models/fs_storage.py:
```diff
@@ -0,0 +1,114 @@
+
+    @api.model
+    def _get_x_accel_redirect_path(self, attachment: IrAttachment):
+        redirect_path = super()._get_x_accel_redirect_path(attachment)
+        fs = self.get_fs_by_code(attachment.fs_storage_code)
+        root_fs = self._get_root_filesystem(fs)
+        if hasattr(root_fs, "s3"):
+            redirect_path = self._get_s3_x_accel_redirect_path(attachment)
+        return redirect_path
```
```suggestion
        fs = self.get_fs_by_code(attachment.fs_storage_code)
        root_fs = self._get_root_filesystem(fs)
        if hasattr(root_fs, "s3"):
            return self._get_s3_x_accel_redirect_path(attachment)
        return super()._get_x_accel_redirect_path(attachment)
```

lmignon committed:
[IMP] fs_attachment: X-Sendfile on S3 storage doesn't require a base url

lmignon commented:
@sbidoul Points that still need improvement:
* [x]  The code still lacks consistency. Odoo uses the term x-sendfile. In the code, I have a mix of method names. Sometimes you find x_accel_redirect, sometimes x_sendfile. I should only keep x_sendfile.
* [x]  Move the implementation of methods  ` _get_x_accel_redirect_path` and `_use_x_sendfile` from fs.storage to ir.attachment (remove this unnecessary delegation of the implemenatiion from ir.attachment to fs.storage).
* [ ] ~~Nice to have: to take advantage of the nginx cache, we should generate signed URLs for s3 with a long lifetime. We should also implement our own URL cache to return the same signed URL for the same object as long as it is not invalidated...~~

lmignon committed:
[IMP] fs_attachment_s3: doc

lmignon committed:
[IMP] fs_attachment, fs_attachement_s3: Code simplification

sbidoul commented on fs_attachment_s3/README.rst:
```diff
@@ -0,0 +1,174 @@
+
+In order to serve files using X-Accel-Redirect, you must ensure that
+your web server is configured to handle these headers correctly. This
+typically involves setting up a location block in your web server
```
```suggestion
typically involves setting up an `internal <https://nginx.org/en/docs/http/ngx_http_core_module.html#internal>`_ location block in your web server
```

lmignon committed:
[IMP] fs_attachment_s3: Improves doc

Co-authored-by: Stéphane Bidoul <stephane.bidoul@acsone.eu>

lmignon committed:
[IMP] fs_attachment, fs_attachment_s3: Naming consistency

Always uses x_sendfile into the code to refer to the x-senfile/x-accel-redirect mechanism

lmignon commented:
@sbidoul We need to adapt our deployements due to the last changes

lmignon committed:
[FIX] fs_attachement_s3: Fix method naming

lmignon commented:
Works in production for the last 2 weeks....

lmignon committed:
[IMP] fs_attachment_s3: Better python dependency

lmignon commented:
/ocabot merge major
